src/compiler/transformer.ts

```diff
@@ -0,0 +1,334 @@
+/// <reference path="visitor.ts" />
+
+/* @internal */
+namespace ts {
+    /**
+     * Transforms an array of SourceFiles by passing them through each transformer.
+     *
+     * @param resolver The emit resolver provided by the checker.
+     * @param host The emit host.
+     * @param sourceFiles An array of source files
+     * @param transforms An array of Transformers.
+     */
+    export function transformFiles(resolver: EmitResolver, host: EmitHost, sourceFiles: SourceFile[], transformers: Transformer[]) {
+        const nodeToGeneratedName: Identifier[] = [];
+        const generatedNameSet: Map<string> = {};
+        const nodeEmitFlags: NodeEmitFlags[] = [];
+        const lexicalEnvironmentVariableDeclarationsStack: VariableDeclaration[][] = [];
+        const lexicalEnvironmentFunctionDeclarationsStack: FunctionDeclaration[][] = [];
+        let lexicalEnvironmentStackOffset = 0;
+        let hoistedVariableDeclarations: VariableDeclaration[];
+        let hoistedFunctionDeclarations: FunctionDeclaration[];
+        let currentSourceFile: SourceFile;
+
+        // The transformation context is provided to each transformer as part of transformer
+        // initialization.
+        const context: TransformationContext = {
+            getCompilerOptions: () => host.getCompilerOptions(),
+            getEmitResolver: () => resolver,
+            getNodeEmitFlags,
+            setNodeEmitFlags,
+            isUniqueName,
+            getGeneratedNameForNode,
+            nodeHasGeneratedName,
+            makeUniqueName,
+            hoistVariableDeclaration,
+            hoistFunctionDeclaration,
+            startLexicalEnvironment,
+            endLexicalEnvironment
+        };
+
+        // Chain together and initialize each transformer.
+        const transformation = chain(...transformers)(context);
+
+        // Transform each source file.
+        return map(sourceFiles, transformSourceFile);
+
+        /**
+         * Transforms a source file.
+         *
+         * @param sourceFile The source file to transform.
+         */
+        function transformSourceFile(sourceFile: SourceFile) {
+            if (isDeclarationFile(sourceFile)) {
+                return sourceFile;
+            }
+
+            currentSourceFile = sourceFile;
+            const visited = transformation(sourceFile);
+            currentSourceFile = undefined;

```

Comment:
Are you asking why I clear `currentSourceFile` after each iteration? I suppose it may be unnecessary.